### ISSUE TITLE
tests fail on Debian Wheezy (7.9) when building App::MechaCPAN 0.25

### ISSUE THREAD

**Issue by pattieja1** [atrodo/App-MechaCPAN#11]
system perl version is 5.14.2

```
cpanm (App::cpanminus) 1.7044 on perl 5.014002 built for x86_64-linux-gnu-thread-multi
Work directory is /root/.cpanm/work/1540476730.30551
You have make /usr/bin/make
You have LWP 6.04
You have /bin/tar: tar (GNU tar) 1.26
Copyright (C) 2011 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
You have /usr/bin/unzip
Searching install () on cpanmetadb ...
--> Working on install
Fetching http://www.cpan.org/authors/id/D/DA/DAGOLDEN/install-0.01.tar.gz
-> OK
Unpacking install-0.01.tar.gz
Entering install-0.01
Checking configure dependencies from META.yml
Checking if you have ExtUtils::MakeMaker 6.58 ... Yes (7.34)
Configuring install-0.01
Running Makefile.PL
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for install
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.34)
Building and testing install-0.01
cp lib/install.pm blib/lib/install.pm
Manifying 1 pod document
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/install.t .. ok
All tests successful.
Files=1, Tests=1,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.01 cusr  0.00 csys =  0.03 CPU)
Result: PASS
Manifying 1 pod document
Installing /usr/local/share/perl/5.14.2/install.pm
Installing /usr/local/man/man3/install.3pm
Appending installation info to /usr/lib/perl/5.14/perllocal.pod
-> OK
Successfully installed install-0.01
Installing /usr/local/share/perl/5.14.2/x86_64-linux-gnu-thread-multi/.meta/install-0.01/MYMETA.json
Installing /usr/local/share/perl/5.14.2/x86_64-linux-gnu-thread-multi/.meta/install-0.01/install.json
Searching App::MechaCPAN () on cpanmetadb ...
--> Working on App::MechaCPAN
Fetching http://www.cpan.org/authors/id/A/AT/ATRODO/App-MechaCPAN-0.25.tar.gz
-> OK
Unpacking App-MechaCPAN-0.25.tar.gz
Entering App-MechaCPAN-0.25
Checking configure dependencies from META.json
Checking if you have ExtUtils::MakeMaker 6.58 ... Yes (7.34)
Configuring App-MechaCPAN-0.25
Running Makefile.PL
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for App::MechaCPAN
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have Test::More 0.96 ... Yes (0.98)
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.34)
Checking if you have CPAN::Meta 0 ... Yes (2.143240)
Building and testing App-MechaCPAN-0.25
cp lib/App/MechaCPAN.pm blib/lib/App/MechaCPAN.pm
cp lib/App/MechaCPAN/Perl.pm blib/lib/App/MechaCPAN/Perl.pm
cp lib/App/MechaCPAN/Install.pm blib/lib/App/MechaCPAN/Install.pm
cp lib/App/MechaCPAN/Deploy.pm blib/lib/App/MechaCPAN/Deploy.pm
"/usr/bin/perl" "-Iblib/arch" "-Iblib/lib" zhuli.PL zhuli
Will include 'MechaCPAN.pm' in 'zhuli'
Will include 'Deploy.pm' in 'zhuli'
Will include 'Install.pm' in 'zhuli'
Will include 'Perl.pm' in 'zhuli'
cp script/mechacpan blib/script/mechacpan
"/usr/bin/perl" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/mechacpan
Manifying 4 pod documents
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t

t/01_sanity.t .............. ok
t/02_cmds.t ................ ok
t/03_run.t ................. ok
t/04_opts.t ................ ok
t/05_re.t .................. ok

t/06_restart.t ............. ok

#   Failed test 'Downloaded tmp file is humane'
#   at t/07_tmpdata.t line 41.
#                   '/tmp/z2ytQWTm8K'
#     doesn't match '(?^msx:^ /tmp/IuC03cIJMx /local/tmp/ (?^msx: mecha_ .*? [.] [\d_]* [.] [^.]* ) $)'

#   Failed test 'Downloaded tmp file includes original filename'
#   at t/07_tmpdata.t line 42.
#                   '/tmp/z2ytQWTm8K'
#     doesn't match '(?^msx: v0\.24\.zip )'

#   Failed test 'Downloaded tmp file was humane'
#   at t/07_tmpdata.t line 103.
#                   '/tmp/WFn2p2rHJx'
#     doesn't match '(?^msx: /tmp/IuC03cIJMx /local/tmp/ (?^msx: mecha_ .*? [.] [\d_]* [.] [^.]* ) )'
# Looks like you failed 3 tests of 35.
t/07_tmpdata.t ............. 
Dubious, test returned 3 (wstat 768, 0x300)
Failed 3/35 subtests 
# logging to '/tmp/POYbG7sAV1/local/logs/mecha_log.20180925_071233.hZz4'...
# Looks like /root/.cpanm/work/1540476730.30551/App-MechaCPAN-0.25/t/../test_dists/FakePerl-5.12.0.tar.gz is perl 5.12.0, assuming that's true
# Fetching perl 5.12.0
# 
# Could not execute 'gzip -dc /root/.cpanm/work/1540476730.30551/App-MechaCPAN-0.25/t/../test_dists/FakePerl-5.12.0.tar.gz | tar xf -'
# PID: 30916	Exit Code: 2		
# 
# 
# gzip: /root/.cpanm/work/1540476730.30551/App-MechaCPAN-0.25/t/../test_dists/FakePerl-5.12.0.tar.gz: not in gzip format
# tar: This does not look like a tar archive
# tar: Exiting with failure status due to previous errors
# 
#  at /root/.cpanm/work/1540476730.30551/App-MechaCPAN-0.25/blib/lib/App/MechaCPAN/Perl.pm line 90
# Configuring perl 5.12.0
# Building perl 5.12.0
# Testing perl 5.12.0
# Installing perl 5.12.0
# Installed perl 5.12.0
t/11_perl.t ................ ok
t/12_perl__get_targz.t ..... ok
t/13_perl_smart_tests.t .... ok
t/14_perl_opts.t ........... ok
t/21_install.t ............. ok
t/22_install__get_targz.t .. ok
t/23_install_array.t ....... ok
t/24_simple_source.t ....... ok
# (BEGIN) Try::Tiny => 
# cwd => /tmp/87WEB16LRN
# Try::Tiny => 
# Try::Tiny => 
# Try::Tiny => 
# Try::Tiny => 
t/25_depend_source.t ....... ok
t/26_avoid_reinstall.t ..... ok
#  - Failed to install 1 modules
#  - Failed to install 1 modules
t/26_keep_processing.t ..... ok
t/26_subseq_prereq.t ....... ok
#  - Failed to install 1 modules
#  - Failed to install 1 modules
#  - Failed to install 1 modules
#  - Failed to install 1 modules
#  - Failed to install 1 modules
t/27_opts.t ................ ok
t/31_cpanfile.t ............ ok
t/32_deploy.t .............. ok
t/33_snapshot.t ............ ok
t/34_deploy_correct.t ...... ok
t/35_deploy_snapshot.t ..... ok
t/author-pod-syntax.t ...... skipped: these tests are for testing by the author
#  - Failed to install 1 modules
#  - Failed to install 1 modules
t/gh3.t .................... ok
t/gh6_git_min_ver.t ........ ok

Test Summary Report
-------------------
t/07_tmpdata.t           (Wstat: 768 Tests: 35 Failed: 3)
  Failed tests:  7-8, 28
  Non-zero exit status: 3
Files=28, Tests=231, 149 wallclock secs ( 0.09 usr  0.04 sys + 66.24 cusr  8.17 csys = 74.54 CPU)
Result: FAIL
Failed 1/28 test programs. 3/231 subtests failed.
make: *** [test_dynamic] Error 255
-> FAIL Installing App::MechaCPAN failed. See /root/.cpanm/work/1540476730.30551/build.log for details. Retry with --force to force install it.
Expiring 3 work directories.
1 distribution installed

```

**Comment by pattieja1** [comment]
specifying --force to the cpanm command did get it to install on Debian Wheezy.  Building on CentOS 7.5.1804 passed the build tests no problem.

**Comment by atrodo** [comment]
Hello.

Excellent! I've seen this error several times on cpantesters, and I've been having a hard time making heads or tails of it.  The error comes down to that MechaCPAN is attempting to download a file to a temporary directory, and getting something completely crazy back:

```
 #   Failed test 'Downloaded tmp file includes original filename'
 #   at t/07_tmpdata.t line 42.
 #                   '/tmp/z2ytQWTm8K'
 #     doesn't match '(?^msx: v0\.24\.zip )'
```

What it should be returning is a file under `/tmp/z2ytQWTm8K`, not that directory itself.

Would it be possible for you to run the following command on that machine and share with me the output? If not, I should be able to get a hold of a Debian 7.9 machine to test on myself.

```
perl -MFile::Temp=tempdir -MApp::MechaCPAN=:go -E '$tmpdir = tempdir( TEMPLATE
 => File::Spec->tmpdir . "/mechacpan_t_XXXXXXXX", CLEANUP => 1 ); $App::MechaCPA
N::PROJ_DIR = $tmpdir; say $tmpdir; say fetch_file("https://get.mechacpan.us/"); say humane_tmpfile("test"); '
```

Basically, I'm testing tempdir, file_fetch and humane_tmpfile to see where it's breaking down, which I think are the only 3 things in play in that failing test.

**Comment by pattieja1** [comment]
```
/tmp/vz4y2pgDYv
Can't call method "scheme" on an undefined value at /usr/local/share/perl/5.14.2/App/MechaCPAN.pm line 548.

```

**Comment by atrodo** [comment]
Sorry about the delay, it's just, I don't know what to make of that. It appears that File::Fetch (through fetch_file) doesn't like "https://get.mechacpan.us/", which it should because in the test output you provided, it was fine with "https://github.com/p5sagit/Try-Tiny/archive/v0.24.zip".

After more research, turns out it's a newer feature of File::Fetch to allow you to fetch bare-urls, so if I had done `https://get.mechacpan.us/v0.25` it would have worked. Getting that error in a fresh 5.14.2 install was helpful and lead me to discover that my previous tests were all messed up.

After that I was able to replicate your error. Turns out that I was feature of File::Temp v0.23, which was released with 5.18, which allows you to pass the template as a single parameter. Previously, you had to specify a hash. Commit 3ee528e47cf10e0e5c849b1d51af1ff4910c9e0a fixes this and will be included with a couple other bugfixes in v0.26

**Comment by pattieja1** [comment]
Glad I could help.  :)

**Comment by pattieja1** [comment]
Is there a branch you want me to test?

**Comment by atrodo** [comment]
Sorry, I hadn't noticed your last comment. master contains the fix so it should work for you on 5.14.

**Comment by atrodo** [comment]
v0.26 contains this commit and I just finished pushing it to cpan.

**Comment by pattieja1** [comment]
I just tested installing App::MechaCPAN 0.26 using Debian Wheezy native perl (5.14.2) and everything went swimmingly.  I'm in the process or re-installing all the perl modules I use to see if anything else comes up but I'm going to say this ticket is resolved. :)
